server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -149,43 +150,48 @@ private Client authorizeClientFromXfccHeader(XfccSourceConfig xfccConfig,
 
     CertificatePrincipal certificatePrincipal =
         new CertificatePrincipal(clientCert.getSubjectDN().toString(),
-            new X509Certificate[] {clientCert});
+            new X509Certificate[] { clientCert });
 
     return authenticateClientFromCertificate(certificatePrincipal);
   }
 
-  private void validateXfccHeaderAllowed(XfccSourceConfig xfccConfig, Principal requestPrincipal) {
+  private void validateXfccHeaderAllowed(XfccSourceConfig xfccConfig, Principal requestPrincipal,
+      ContainerRequest containerRequest) {
     if (clientAuthConfig == null || clientAuthConfig.xfccConfigs() == null) {
       throw new NotAuthorizedException(
           format(
               "not configured to handle requests with %s header set; set 'xfcc' field in config",
               XFCC_HEADER_NAME));
     }
 
+    Optional<String> spiffeIdHeader = Optional.ofNullable(xfccConfig.spiffeIdHeader());
+    Optional<URI> callerSpiffeId = spiffeIdHeader.isPresent()
+        ? ClientAuthenticator.getSpiffeIdFromHeader(containerRequest, spiffeIdHeader.get())
+        : Optional.empty();
+
     // All XFCC traffic must be checked against allowlists; all connections that set the XFCC
     // header must send identifiers for the requester, as well as the XFCC certificate
     Optional<String> requestName = ClientAuthenticator.getClientName(requestPrincipal);
-    Optional<URI> requestSpiffeId = ClientAuthenticator.getSpiffeId(requestPrincipal);
+    Optional<URI> requestSpiffeId = callerSpiffeId.isPresent()
+        ? callerSpiffeId : ClientAuthenticator.getSpiffeId(requestPrincipal);
```

Comment:
This should always be `ClientAuthenticator.getSpiffeId(requestPrincipal)`.

There are two entities involved in the process of handling a proxied connection to Keywhiz: the proxy and the actual client whose request was proxied. This code uses the "request" prefix to indicate data related to the proxy, because that data comes from the mTLS request that Keywhiz received directly. The requester/proxy should _not_ be identified based on a header.

This function only checks that the entity which connected to Keywhiz is allowed to do so, so the forwarded "caller spiffe ID" shouldn't be used here.

The code that needs to be updated is the "extract client information from XFCC header" section starting in line 144--that's where the client whose request is being proxied is identified. Currently there's only one branch there because the only way that a client can be identified is from the XFCC header, but we should add a new case for a client identified by the caller-spiffe-ID header.

(Please feel free to add comments describing the logic--it's more confusing than I initially realized when writing it.)

Reply:
Thanks for the feedback and the pointer. So for the caller authentication can I extract the Spiffe Id from the header and check if it's in the `allowedCallerSpiffeId`? Or do you think I should create a principal extracted from the caller Spiffe Id and reuse `authenticateClientFromCertificate`?

Reply:
Handcrafting a principal is pretty tricky iirc, so I'd weakly recommend not trying to reuse `authenticateClientFromCertificate`--but you can probably refactor ClientAuthenticator to share code between `authenticateClientFromCertificate` and `authenticateClientFromSpiffeId` (or whatever the new method's called). But it's definitely worth checking whether you can create a principal with the caller Spiffe ID and use it in a (renamed) `authenticateClientFromCertificate` method.

Reply:
@jbpeirce I made some changes so that now we authorize the requests from a custom header starting in line 144. Thank you again for the pointer. I didn't fully understand the auth process in my previous iterations. Let me know if this looks good to you.